design_documents/cairo_deployment_scripts.md

```diff
@@ -0,0 +1,335 @@
+
+### error handling
+If the transaction fails, we should put all the relevant info about it to the state file, output log information and stop
+script execution.
+If the script fails (for any reason that is not connected to transactions - eg rpc node down), we should output relevant
+log information.
+
+In case of failed transaction, in the "output" field in state file we could include an error message. After fixing the issue
+we should allow users to just re-run the script - all the previous (succeeded) transactions should not be replayed, the
+erroneous transaction should be retried and its output should be put into state file.
+
+### idempotency
+At the later stages we will want to have the script to be able to track the on chain state using a state file, which would 
+in turn allow the script to be called multiple times without any modifications, to ensure desired state. To achieve that
+we must ensure idempotency of calls to the network.
+
+The proposed solution would be to use an external json file, that holds hashes of transactions that were already performed.
+While executing a command we could first check this file to see if given transaction exists/succeeded, and based on this
+either perform and an action (make a transaction) or skip it and proceed to the next thing in script.
+
+To achieve this there are a few possible solutions:
+- adjust our existing cast subcommand functions to handle checking the file
+  - pros: 
+    - just import a subcommand function and use it without a fuss
+    - easy to implement for cast subcommands functions
+  - cons:
+    - we will have to modify currently existing functions to enable state checking
+    - this solution does not work for dispatcher functions
+
+- just try to execute functions, parse the output and compare it with the state file
+  - pros:
+    - no modifications needed for subcommands functions
+    - just import a function and use it
+  - cons:
+    - kind of brute force solution - we should not have to talk to chain since we have all the necessary info in local file
+    - executing dispatcher functions can cause us all kinds of trouble (eg - someone could invoke 'send_eth' function multiple times, where in fact they only wanted to do it once)
+    - not every cast subcommand is idempotent
+
+- implement a higher order function (or function wrapper) to conditionally execute functions based on the information in state file
+  - pros:
+    - no modifications needed to cast subcommands functions
+    - works with dispatcher functions just fine
+    - added flexibility - user can decide which function should be replayed each time, and which shouldn't
+  - cons:
+    - boilerplate code - wrapping each function call will be ugly
+    - no option to just annotate main function (procedural macros could achieve this, but they do some compile-time modifications)
+    - potentially more complex
```

Comment:
I think cast commands, when run from the command line should just behave as they did - without checking the state. But it's just my opinion. 

If we followed this approach, I think the solution with wrappers on the basic functions or macros seems to be a preferred solution.

Reply:
Also, if it possible at all, these functions could have the same names as the command line commands, but have this slightly different behavior where they check the state.

Reply:
> I think cast commands, when run from the command line should just behave as they did - without checking the state. But it's just my opinion.

Yup, that's the plan - these will be the same functions. The logic would be there in already present cast functions, but it will only be used in deployment scripts. Probably some input parameter should suffice.